Controller/ExportController.php

```diff
@@ -11,145 +11,152 @@
-                        // insert header row?
-                        if (!$headersWritten) {
-                            $headers = array_keys($contact);
-                            // make sure to add dnc headers since 1st record may not have them
-                            if (empty($dnc)) {
-                                $dnc_headers = ['dnc_date_added', 'dnc_reason', 'dnc_channel', 'dnc_channel_id', 'dnc_comments'];
-                                $headers = array_merge($headers, $dnc_headers);
-                            }
-                            fputcsv($handle, $headers);
-                            $headersWritten = true;
+
+        list($dateFrom, $dateTo)  = $this->convertDateParams($dateFrom, $dateTo);
+        $contactIds               = $this->getCampaignLeadIdsForExport($campaign->getId(), $dateFrom, $dateTo);
+
+        // adjust $size for memory vs. speed
+        $batches = array_chunk($contactIds, 100);
+
+        //testing
+        /** @var OverrideLeadRepository $contactRepo */
+        $contactRepo = $this->getModel('lead')->getRepository();
+
+        $fileName = sprintf('ContactsExportFrom%s.csv', str_replace(' ', '', $campaign->getName()));
+
+        $response = new StreamedResponse(
+            function () use ($contactRepo, $batches) {
+                ini_set('max_execution_time', 0);
+                $handle = fopen('php://output', 'w');
+
+                $fieldNames = [];
+                foreach ($batches as $batch) {
+                    //$leads = $contactRepo->getEntitiesWithCustomFields('lead', ['ids' => $batch]);
+                    $leads = $contactRepo->getEntities(['ids' => $batch, '', 'withTotalCounts'=> 0, 'withChannelRules' => 1, 'ignore_paginator' => 1]);
+                    /**
+                     * @var int
+                     * @var Lead $lead
+                     */
+                    foreach ($leads as $id => $lead) {
+                        if (empty($fieldNames)) {
+                            $fields = $lead->getFields(true);
+                            $columnNames = array_map(function ($f) { return $f['label']; }, $fields);
+                            $columnNames = array_merge(['Id'], $columnNames);
+                            fputcsv($handle, $columnNames);
+                            $fieldNames = array_map(function ($f) { return $f['alias']; }, $fields);
                         }
-                        fputcsv($handle, array_values($contact));
+                        $values = [$id];
+                        foreach ($fieldNames as $fieldName) {
+                            $values[] = $lead->getFieldValue($fieldName);
+                        }
+                        fputcsv($handle, $values);
                     }
-                    $start += $params['limit'];
-                    $this->container->get('doctrine.orm.entity_manager')->clear();
-                    gc_enable();
```

Comment:
Is gc_collection not needed? Do you not exceed memory limits after a few hundred batches if we do not clear the entity manager? Are the leads no longer in memory?

Reply:
not needed. garbage is not part of the memory limit associated with this process. By looping over the results and reassinging to the same variable, memory limits are conserved. I used the laravel StreamedResponse as a model and this issue was addressed